article/app/model/dotcomponents/DotcomponentsDataModel.scala

```diff
@@ -218,18 +219,50 @@ object DotcomponentsDataModel {
 
   val VERSION = 1
 
-  def fromArticle(articlePage: ArticlePage, request: RequestHeader): DotcomponentsDataModel = {
+  def fromArticle(articlePage: ArticlePage, request: RequestHeader, blocks: APIBlocks): DotcomponentsDataModel = {
 
     val article = articlePage.article
 
-    val bodyBlocks: List[Block] = article.blocks match {
-      case Some(bs) => bs.body.map(bb => Block(bb.bodyHtml, bb.dotcomponentsPageElements.toList)).toList
-      case None => List()
+    // TODO this logic is duplicated from the cleaners, can we consolidate?
+    val shouldAddAffiliateLinks = AffiliateLinksCleaner.shouldAddAffiliateLinks(
+      switchedOn = Switches.AffiliateLinks.isSwitchedOn,
+      section = article.metadata.sectionId,
+      showAffiliateLinks =  article.content.fields.showAffiliateLinks,
+      supportedSections = affiliatelinks.affiliateLinkSections,
+      defaultOffTags = affiliatelinks.defaultOffTags,
+      alwaysOffTags = affiliatelinks.alwaysOffTags,
+      tagPaths = article.content.tags.tags.map(_.id)
+    )
+
+    def blocksToPageElements(elements: Seq[ClientBlockElement], affiliateLinks: Boolean): List[PageElement] = {
+      elements.toList.flatMap(el => PageElement.make(
+        element = el,
+        addAffiliateLinks = affiliateLinks,
+        pageUrl = request.uri
+      ))
     }
 
-    val mainBlock: Option[Block] = article.blocks.flatMap(
-      _.main.map(bb=>Block(bb.bodyHtml, bb.dotcomponentsPageElements.toList))
-    )
+    def addDisclaimer(elems: List[PageElement]): List[PageElement] = {
+      val hasLinks = elems.exists({
+        case TextBlockElement(html) => AffiliateLinksCleaner.stringContainsAffiliateableLinks(html)
+        case _ => false
+      })
+
+      if (hasLinks) {
+        elems :+ DisclaimerBlockElement(affiliateLinksDisclaimer("article").body)
+      } else {
+        elems
+      }
+    }
+
+    val bodyBlocks: List[Block] = {
+      val bodyBlocks = blocks.body.getOrElse(Nil)
+      bodyBlocks.map(block => Block("", blocksToPageElements(block.elements, shouldAddAffiliateLinks))).toList
```

Comment:
Note to self and others, I've set the body here to the empty string on the assumption we don't need this in DC, but maybe I should see if I can do this better?

Reply:
(This is now outdated.)